test/mako/config/benchmark.go

```diff
@@ -38,13 +41,38 @@ func MustGetBenchmark() (*string, *string) {
 	return benchmarkKey, benchmarkName
 }
 
+type BenchmarkEnv struct {
+	DevKey   string `envconfig:"DEV_BENCHMARK_KEY"`
+	DevName  string `envconfig:"DEV_BENCHMARK_NAME"`
+	ProdKey  string `envconfig:"PROD_BENCHMARK_KEY"`
+	ProdName string `envconfig:"PROD_BENCHMARK_NAME"`
+}
+
 // getBenchmark fetches the appropriate benchmark_key for this configured environment.
 func getBenchmark() (*string, *string, error) {
 	// Figure out what environment we're running in from the Mako configmap.
 	env, err := getEnvironment()
 	if err != nil {
 		return nil, nil, err
 	}
+	// If pod has benchmark related env
+	benchmarkEnv := BenchmarkEnv{}
+	if err := envconfig.Process("", &benchmarkEnv); err != nil {
+		return nil, nil, err
+	}
+	if !reflect.DeepEqual(benchmarkEnv, BenchmarkEnv{}) {
+		if env == "dev" {
+			if benchmarkEnv.DevKey != "" && benchmarkEnv.DevName != "" {
+				return &benchmarkEnv.DevKey, &benchmarkEnv.DevName, nil
+			}
+			return nil, nil, errors.New("failed to get benchmark config from environment variables")
```

Comment:
I'm not a fan of this. It seems like you could end up with the following:

(empty config file)

> Crashes if only PROD_BENCHMARK_NAME and PROD_BENCHMARK_KEY are set.

> Continues if only DEV_BENCHMARK_NAME and DEV_BENCHMARK_KEY are set.

> Continues if both DEV_BENCHMARK_NAME/KEY and PROD_BENCHMARK_NAME/KEY are set, but always uses the "dev" value.


My $0.02 would be to move this from files in the container via `kodata` to environment variables in most/all cases.

Reply:
Additionally, if you use only a single set of env vars, you could name them "Name" and "Key", and use "benchmark" as the first argument to `envconfig.Process()`.

Reply:
I think that would make the error handling here nicer, too.

Reply:
Would it be possible to give me more explanation for the single set of environment vars in container? 

I use the `env` variable got from configMap to check whether I need to use `DEV_BENCHMARK_*` values or `PROD_BENCHMARK_*` values in this part of code. It will go for the following scenario:
1. `env` is not set in configMap (error is handled in `env, err := getEnvironment())` at the beginning of this function)
2. `env` is `dev` and there is at least one `*_BENCHMARK_*` value set in the container
    	(1).`DEV_BENCHMARK_NAME` and `DEV_BENCHMARK_KEY` are set, return `DEV_BENCHMARK_NAME/KEY` values
        (2). Only one value from `DEV_BENCHMARK_NAME` and `DEV_BENCHMARK_KEY` is set, crash
        (3).`DEV_BENCHMARK_NAME/KEY` are not set, crash
3. `env` is `prod` and there is at least one `*_BENCHMARK_*` value set in the container
      ...logic is the same as `env` is `dev`
4. `env` is other value (continue, error is handled in the later part of this function)

Reply:
IMO, we should either use environment variables or file contents; mixing the two is a bit strange.

If you want to use the ConfigMap with env vars, you can reflect ConfigMap data into environment vars:

https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-container-environment-variables-using-configmap-data

Alternatively, you could extract the name and key from the config file if present, rather than extracting it from env vars.